<a href="https://colab.research.google.com/github/ritubhatiagmail/python/blob/master/Seshat_Duplicate_Algo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Imoprting libraries

In [ ]:
!pip install -U sentence-transformers
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('bert-base-nli-mean-tokens')
!pip install ipython-autotime
%load_ext autotime
!sudo apt install tesseract-ocr
!pip install pytesseract
from nltk.corpus import stopwords 
from nltk.tokenize import word_tokenize 
import nltk
nltk.download('all')

     |████████████████████████████████| 71kB 3.2MB/s 
     |████████████████████████████████| 778kB 5.9MB/s 
     |████████████████████████████████| 1.1MB 47.2MB/s 
     |████████████████████████████████| 890kB 38.2MB/s 
     |████████████████████████████████| 3.0MB 40.3MB/s 
  Created wheel for sentence-transformers: filename=sentence_transformers-0.3.0-cp36-none-any.whl size=86756 sha256=0bd70601202f43ee90dde72f86ac919bc179d2c5d858e2d4fabe1d07c31dcd0e
  Stored in directory: /root/.cache/pip/wheels/b8/23/85/85d6a9a6c68f0625a1ecdaad903bb0a78df058c10cf74f9de4
  Created wheel for sacremoses: filename=sacremoses-0.0.43-cp36-none-any.whl size=893260 sha256=fe72b0c13c5e113403b4143434a47f52d8abe709d5d9de8ff8eb4fc0be56ff91
  Stored in directory: /root/.cache/pip/wheels/29/3c/fd/7ce5c3f0666dab31a50123635e6fb5e19ceb42ce38d4e58f45
Successfully built sentence-transformers sacremoses


100%|██████████| 405M/405M [00:45<00:00, 8.98MB/s]


  Created wheel for ipython-autotime: filename=ipython_autotime-0.1-cp36-none-any.whl size=1832 sha256=2b420dd7b6a0bdc24c202f50d1081c9170a0c57db9edd2fe06b33c54b092bddb
  Stored in directory: /root/.cache/pip/wheels/d2/df/81/2db1e54bc91002cec40334629bc39cfa86dff540b304ebcd6e
Successfully built ipython-autotime
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following package was automatically installed and is no longer required:
  libnvidia-common-440
Use 'sudo apt autoremove' to remove it.
The following additional packages will be installed:
  tesseract-ocr-eng tesseract-ocr-osd
The following NEW packages will be installed:
  tesseract-ocr tesseract-ocr-eng tesseract-ocr-osd
0 upgraded, 3 newly installed, 0 to remove and 35 not upgraded.
Need to get 4,795 kB of archives.
After this operation, 15.8 MB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 tesseract-ocr-eng all 4.00~git24-0e0

True

In [ ]:
import spacy
import string
import re
import pytesseract
import shutil
import os
import random
try:
    from PIL import Image
except ImportError:
    import Image
from scipy import signal
import cv2
import numpy as np
import cv2
from google.colab.patches import cv2_imshow
import matplotlib.pyplot as plt
from PIL import Image
import imageio
import pandas as pd
from scipy import spatial

time: 635 ms


# Getting all files

In [ ]:
files_P, files_P_op, files_P_sol = [], [], []
files_B, files_B_op, files_B_sol = [] , [], []
files_M, files_M_op, files_M_sol = [] , [], []
files_C, files_C_op, files_C_sol = [] ,[], []
for subdir, dirs, files in os.walk(r'/content/drive/My Drive/Duplicate_Algo_280620'):
    files.sort()
    for filename in files:
       if filename.endswith(".GIF"):
        if ((filename[8] == '0') & (filename[9] == '0')):
          if filename.startswith("P"):
            files_P.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("B"):
            files_B.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("M"):
            files_M.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("C"):
            files_C.append(os.path.abspath(os.path.join(subdir, filename)))
        if ((filename[8] == '0') & (filename[9] == '1')):
          if filename.startswith("P"):
            files_P_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("B"):
            files_B_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("M"):
            files_M_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("C"):
            files_C_op.append(os.path.abspath(os.path.join(subdir, filename)))
        if ((filename[8] == '0') & (filename[9] == '2')):
          if filename.startswith("P"):
            files_P_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("B"):
            files_B_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("M"):
            files_M_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("C"):
            files_C_op.append(os.path.abspath(os.path.join(subdir, filename)))   
        if ((filename[8] == '0') & (filename[9] == '3')):
          if filename.startswith("P"):
            files_P_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("B"):
            files_B_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("M"):
            files_M_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("C"):
            files_C_op.append(os.path.abspath(os.path.join(subdir, filename)))
        if ((filename[8] == '0') & (filename[9] == '4')):
          if filename.startswith("P"):
            files_P_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("B"):
            files_B_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("M"):
            files_M_op.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("C"):
            files_C_op.append(os.path.abspath(os.path.join(subdir, filename)))                                         
        if ((filename[8] == '9') & (filename[9] == '9')):
          if filename.startswith("P"):
            files_P_sol.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("B"):
            files_B_sol.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("M"):
            files_M_sol.append(os.path.abspath(os.path.join(subdir, filename)))
          elif filename.startswith("C"):
            files_C_sol.append(os.path.abspath(os.path.join(subdir, filename)))              

time: 10.4 s


In [ ]:
print(len(files_P), len(files_B), len(files_M), len(files_C) )
print(len(files_P_op), len(files_B_op), len(files_M_op), len(files_C_op) )
print(len(files_P_sol), len(files_B_sol), len(files_M_sol), len(files_C_sol))

20 20 19 20
76 80 76 76
19 0 20 20
time: 3.01 ms


In [ ]:
files_P[0][-12:-7]

'08354'

time: 14.1 ms


# Compare image Algo

In [ ]:
similar_df_Pi = pd.DataFrame(columns=['Id_Q1', 'Id_Q2', 'Similarity_Ques'])
import cv2
for i in range(len(files_P)):
  for j in range(i+1, len(files_P)):
    gif = imageio.get_reader(files_P[i])
    gif2 = imageio.get_reader(files_P[j])
    for frame in gif:
      image_1 = frame[:,:,0]
    for frame in gif2:
      image_2 = frame[:,:,0]
    first_image_hist = cv2.calcHist([image_1], [0], None, [256], [0, 256])
    second_image_hist = cv2.calcHist([image_2], [0], None, [256], [0, 256])

    img_hist_diff = cv2.compareHist(first_image_hist, second_image_hist, cv2.HISTCMP_BHATTACHARYYA)
    img_template_probability_match = cv2.matchTemplate(first_image_hist, second_image_hist, cv2.TM_CCOEFF_NORMED)[0][0]
    img_template_diff = 1 - img_template_probability_match

    # taking only 10% of histogram diff, since it's less accurate than template method
    commutative_image_diff = (img_hist_diff / 10) + img_template_diff
    similar_df_Pi = similar_df_Pi.append({'Id_Q1':files_P[i][-14:-7], 'Id_Q2':files_P[j][-14:-7], 'Similarity_Ques':commutative_image_diff*1000}, ignore_index=True)
    #print( commutative_image_diff*10000)

similar_df_Pi = similar_df_Pi.sort_values(by=['Similarity_Ques'])
similar_df_Pi.reset_index(drop=True, inplace=True)
similar_df_Pi = similar_df_Pi.round(2)
similar_df_Pi2 = similar_df_Pi.iloc[:20,:]
similar_df_Pi2

Id_Q1    Id_Q2  Similarity_Ques
0   P-03149  P-08402             0.30
1   P-08354  P-08394             0.54
2   P-50273  P-50193             0.80
3   P-08394  P-08402             0.83
4   P-50273  P-50040             0.84
5   P-50193  P-50293             0.84
6   P-08394  P-03149             0.92
7   P-50193  P-04563             0.95
8   P-08354  P-08402             0.95
9   P-00202  P-04899             0.96
10  P-04406  P-22405             0.99
11  P-08354  P-03149             1.07
12  P-50273  P-22405             1.08
13  P-22405  P-50293             1.18
14  P-50273  P-50293             1.19
15  P-01118  P-02504             1.26
16  P-04406  P-02504             1.33
17  P-00202  P-07970             1.39
18  P-03149  P-04563             1.40
19  P-50193  P-50040             1.42

time: 10.4 s


Change the image path to the desired path to save result images

In [ ]:
image_path = '/content/drive/My Drive/Duplicate_Algo_280620/Results3(image)'
count=1
for j in range(len(similar_df_Pi2)):
  for i in range(len(files_P)):
    if files_P[i][-14:-7] == similar_df_Pi2.iloc[j,0]:
      A = plt.imread(files_P[i])
      #print(files_P[i][-14:-7])
    if files_P[i][-14:-7] == similar_df_Pi2.iloc[j,1]:
      #print(files_P[i][-14:-7])
      B = plt.imread(files_P[i])
  #print("Other")
  width = A.shape[1]
  height = A.shape[0]
  dsize = (width, height)
  output = cv2.resize(B, dsize)
  C = np.hstack((A, output))
  #cv2_imshow(C)
  cv2.imwrite(os.path.join(image_path,('P' + str(count) + '_' +  (str(similar_df_Pi2.iloc[j,2]))[:5]+".jpg")),C)
  count+=1

time: 153 ms


# Ocr detection

In [ ]:
def simi_fn2(X,Y):
  dataSetI = X
  dataSetII = Y
  result = 1 - spatial.distance.cosine(dataSetI, dataSetII)
  return (result)

time: 1.53 ms


In [ ]:
df = pd.DataFrame(columns=['Id', 'Questions', 'Options', 'Solution'])
for filename in files_P:
  temp=[]
  for i in range(len(files_P_op)):
    if files_P_op[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_P_op[i])
      for frame in gif:
          frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
          info = pytesseract.image_to_string(frame)
          temp.append(info)
  temp.sort()
  
  temp2 = ' '.join(temp)
  #temp2 = temp2.replace(" ","")
  temp3=[]
  for i in range(len(files_P_sol)):
    if files_P_sol[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_P_sol[i])
      for frame in gif:
          frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
          info = pytesseract.image_to_string(frame)
          temp3.append(info)
  temp3.sort()
  temp4 = ' '.join(temp3)
  #temp4 = temp4.replace(" ","")
  gif = imageio.get_reader(filename)
  for frame in gif:
    frame = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    info = pytesseract.image_to_string(frame)
    #info = info.replace(" ","")
  df = df.append({'Id':filename[-14:-7],'Questions':info, 'Options':temp2, 'Solution':temp4},ignore_index=True)
df['Questions'].astype(str)

0     A bullet of mass m and speed v hits a pendulum...
1     A bullet of mass m and speed v hits a pendulum...
2     If the condenser of capacitance 2F shown in th...
3     In the condenser show in the circuit is charge...
4     A plane electromagnetic wave of frequency 25MH...
5     A plane electromagnetic wave of frequency 50 M...
6          X-rays can not penetrate through a sheet of:
7                              X-rays do not penetrate:
8     Two particles X and Y having equal charges, af...
9     Two particles A and B having equal charges aft...
10    A small hole is made at the bottom of a symmet...
11    A small hole is made at the bottom of a symmet...
12    If potential difference of 10,000 volt is appl...
13    A potential difference of 20 kV is applied acr...
14    A inductor 20 mH a capacitor 100 uF and a resi...
15    A inductor 20 mH a capacitor 100 uF and a resi...
16    An inductor 20 mH, a capacitor 50 pF and a res...
17    When the hydrogen atom emits a photon in g

time: 1min 10s


In [ ]:
df["Questions"] = df["Questions"].str.lower()
df["Options"] = df["Options"].str.lower()
df["Solution"] = df["Solution"].str.lower()

time: 4.54 ms


In [ ]:
nltk.download('stopwords')
STOPWORDS = set(stopwords.words('english'))
def remove_stopwords(text):
    """custom function to remove the stopwords"""
    return " ".join([word for word in str(text).split() if word not in STOPWORDS])

df["Questions"] = df["Questions"].apply(lambda text: remove_stopwords(text))
df["Options"] = df["Options"].apply(lambda text: remove_stopwords(text))
df["Solution"] = df["Solution"].apply(lambda text: remove_stopwords(text))

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
time: 15.7 ms


In [ ]:
questions = df['Questions'].to_list()
options = df['Options'].to_list()
solutions = df['Solution'].to_list()

time: 2.42 ms


In [ ]:
questions_embeddings = model.encode(questions)
options_embeddings = model.encode(options)
solutions_embeddings = model.encode(solutions)

time: 11.2 s


In [ ]:
similar_df_P = pd.DataFrame(columns=['Id_Q1', 'Id_Q2', 'Similarity_Ques', 'Similarity_Opt', 'Similarity_Sol'])
for i in range(df.shape[0]):
  for j in range(i+1, df.shape[0]):
    score1 = simi_fn2(questions_embeddings[i], questions_embeddings[j])
    score2 = simi_fn2(options_embeddings[i], options_embeddings[j])
    score3 = simi_fn2(solutions_embeddings[i], solutions_embeddings[j])
    if score1 > 0.7:
         similar_df_P = similar_df_P.append({'Id_Q1':df.iloc[i,0], 'Id_Q2':df.iloc[j,0], 'Similarity_Ques':score1, 'Similarity_Opt':score2, 'Similarity_Sol':score3}, ignore_index=True)
print(similar_df_P)  


      Id_Q1    Id_Q2  Similarity_Ques  Similarity_Opt  Similarity_Sol
0   P-08354  P-08394         0.998420        0.946974        0.820198
1   P-00202  P-04899         0.906202        0.633348        0.473455
2   P-00202  P-50193         0.720889        0.503206        0.856458
3   P-00202  P-50293         0.730348        0.545546        0.900034
4   P-00202  P-50040         0.739488        0.482057        0.614653
5   P-04406  P-50273         0.890714        0.828299        0.367926
6   P-04406  P-01118         0.714858        0.456922        0.234293
7   P-02491  P-02546         0.954623        0.448601        0.994799
8   P-01118  P-22405         0.965802        0.731623        0.579725
9   P-03149  P-08402         0.986351        1.000000        0.951037
10  P-02551  P-02504         0.766654        0.678473        0.662023
11  P-02551  P-50193         0.755492        0.645219        0.681305
12  P-02551  P-50293         0.752156        0.729623        0.760150
13  P-02551  P-50040

1.   21 dup, 144 ms, 42 for ocr(with 4 channel) 
2.   21 dup, 151 ms, 41.8 for ocr(With4) without spaces
3.   21 dup, 147, 42.5 1 channel
2.   21 dup, 127, 38.3, 1 channel with space
1.   21 dup, 151, 38.3+4, lowercase
1.   16 dup, removed punctuations
2.   22 dup, 150, 38.2+4+13, lowercase, removed stopwords
1.   18 dup, 167ms, removed 10 most common words
2.   22 dup,167 ms, removed 2 most common words










In [ ]:
image_path = '/content/drive/My Drive/Duplicate_Algo_280620/Results2'
count=1
for j in range(len(similar_df_P)):
  for i in range(len(files_P)):
    if files_P[i][-14:-7] == similar_df_P.iloc[j,0]:
      A = plt.imread(files_P[i])
    if files_P[i][-14:-7] == similar_df_P.iloc[j,1]:
      B = plt.imread(files_P[i])
  width = A.shape[1]
  height = A.shape[0]
  dsize = (width, height)
  output = cv2.resize(B, dsize)
  C = np.hstack((A, output))
  #cv2_imshow(C)
  cv2.imwrite(os.path.join(image_path,('P' + str(count) + '_' +  (str(similar_df_P.iloc[j,2]))[:5]+".jpg")),C)
  count+=1

time: 137 ms


# Ocr 2

In [ ]:
similar_df_P['combine']= similar_df_P.Id_Q1 + ',' + similar_df_P.Id_Q2
similar_df_P

Id_Q1    Id_Q2  ...  Similarity_Sol          combine
0   P-08354  P-08394  ...        0.820198  P-08354,P-08394
1   P-00202  P-04899  ...        0.473455  P-00202,P-04899
2   P-00202  P-50193  ...        0.856458  P-00202,P-50193
3   P-00202  P-50293  ...        0.900034  P-00202,P-50293
4   P-00202  P-50040  ...        0.614653  P-00202,P-50040
5   P-04406  P-50273  ...        0.367926  P-04406,P-50273
6   P-04406  P-01118  ...        0.234293  P-04406,P-01118
7   P-02491  P-02546  ...        0.994799  P-02491,P-02546
8   P-01118  P-22405  ...        0.579725  P-01118,P-22405
9   P-03149  P-08402  ...        0.951037  P-03149,P-08402
10  P-02551  P-02504  ...        0.662023  P-02551,P-02504
11  P-02551  P-50193  ...        0.681305  P-02551,P-50193
12  P-02551  P-50293  ...        0.760150  P-02551,P-50293
13  P-02551  P-50040  ...        0.454667  P-02551,P-50040
14  P-02504  P-50193  ...        0.754868  P-02504,P-50193
15  P-02504  P-50293  ...        0.752085  P-02504,P-50293
16  P-02504  P-50040  ...        0.541048  P-02504,P-50040
17  P-50193  P-50293  ...        0.954379  P-50193,P-50293
18  P-50193  P-50040  ...        0.727407  P-50193,P-50040
19  P-50293  P-50040  ...        0.675074  P-50293,P-50040
20  P-50293  P-04563  ...        0.756557  P-50293,P-04563
21  P-07970  P-21468  ...        0.147846  P-07970,P-21468

[22 rows x 6 columns]

time: 40.3 ms


In [ ]:
def simi_fn(X,Y):
  X_list = word_tokenize(X)  
  Y_list = word_tokenize(Y) 
  sw = stopwords.words('english')  
  l1 =[];l2 =[] 
  X_set = {w for w in X_list if not w in sw}  
  Y_set = {w for w in Y_list if not w in sw} 
  rvector = X_set.union(Y_set)  
  for w in rvector: 
      if w in X_set: l1.append(1) 
      else: l1.append(0) 
      if w in Y_set: l2.append(1) 
      else: l2.append(0) 
  c = 0
  for i in range(len(rvector)): 
          c+= l1[i]*l2[i] 
  if (((sum(l1)*sum(l2))**0.5) ) == 0:
    return 0
  else:
    cosine = c / float((sum(l1)*sum(l2))**0.5) 
    return cosine

time: 15.2 ms


In [ ]:
similar_df_P2 = pd.DataFrame(columns=['Id_Q1', 'Id_Q2', 'Similarity_Ques', 'Similarity_Opt', 'Similarity_Sol'])
for k in range(df.shape[0]):
  for l in range(k+1, df.shape[0]):
     X =df.iloc[k,1]
     Y =df.iloc[l,1]
     Z =df.iloc[k,2]
     W =df.iloc[l,2]
     S1 = df.iloc[k,3]
     S2 = df.iloc[l,3]
     score1 = simi_fn(X,Y)
     score2 = simi_fn(Z,W)
     score3 = simi_fn(S1, S2)
    #print("similarity: ", cosine) 
     if score1 > 0.7:
      similar_df_P2 = similar_df_P2.append({'Id_Q1':df.iloc[k,0], 'Id_Q2':df.iloc[l,0], 'Similarity_Ques':score1, 'Similarity_Opt':score2, 'Similarity_Sol':score3}, ignore_index=True)
#print(similar_df_P2)  
similar_df_P2['combine']= similar_df_P2.Id_Q1 + ',' + similar_df_P2.Id_Q2
similar_df_P2

Id_Q1    Id_Q2  ...  Similarity_Sol          combine
0  P-08354  P-08394  ...        0.804132  P-08354,P-08394
1  P-02491  P-02546  ...        0.935414  P-02491,P-02546
2  P-01118  P-22405  ...        0.293470  P-01118,P-22405
3  P-03149  P-08402  ...        0.759072  P-03149,P-08402
4  P-50193  P-50293  ...        0.344124  P-50193,P-50293
5  P-50193  P-50040  ...        0.216777  P-50193,P-50040
6  P-50293  P-50040  ...        0.151186  P-50293,P-50040
7  P-07970  P-21468  ...        0.000000  P-07970,P-21468

[8 rows x 6 columns]

time: 483 ms


In [ ]:
mergedStuff = pd.merge(similar_df_P, similar_df_P2, on=['combine'], how='inner')
print(mergedStuff.shape[0])
print(similar_df_P.shape[0])
print(similar_df_P2.shape[0])

8
22
8
time: 26.7 ms


In [ ]:
df = pd.DataFrame(columns=['Id', 'Questions', 'Options', 'Solution'])
for filename in files_M:
  temp=[]
  for i in range(len(files_M_op)):
    if files_M_op[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_M_op[i])
      for frame in gif:
          info = pytesseract.image_to_string(frame)
          temp.append(info)
  temp.sort()
  temp2 = ' '.join(temp)
  temp3=[]
  for i in range(len(files_M_sol)):
    if files_M_sol[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_M_sol[i])
      for frame in gif:
          info = pytesseract.image_to_string(frame)
          temp3.append(info)
  temp3.sort()
  temp4 = ' '.join(temp3)
  gif = imageio.get_reader(filename)
  for frame in gif:
    info = pytesseract.image_to_string(frame)
  df = df.append({'Id':filename[-14:-7],'Questions':info, 'Options':temp2, 'Solution':temp4},ignore_index=True)
df['Questions'].astype(str)
print(df.head())
similar_df_M = pd.DataFrame(columns=['Id_Q1', 'Id_Q2', 'Similarity_Ques', 'Similarity_Opt', 'Similarity_Sol'])
for k in range(df.shape[0]):
  for l in range(k+1, df.shape[0]):
     X =df.iloc[k,1]
     Y =df.iloc[l,1]
     Z =df.iloc[k,2]
     W =df.iloc[l,2]
     S1 = df.iloc[k,3]
     S2 = df.iloc[l,3]
     score1 = simi_fn(X,Y)
     score2 = simi_fn(Z,W)
     score3 = simi_fn(S1, S2)
    #print("similarity: ", cosine) 
     if score1 > 0.7:
      similar_df_M = similar_df_M.append({'Id_Q1':df.iloc[k,0], 'Id_Q2':df.iloc[l,0], 'Similarity_Ques':score1, 'Similarity_Opt':score2, 'Similarity_Sol':score3}, ignore_index=True)
print(similar_df_M) 

        Id  ...                                           Solution
0  M-09045  ...  Solution:\nf(x) =x? +bx? +ex4+d,0<b? <c\nf(x) ...
1  M-07594  ...  Solution:\n\nGiven f(x) =x + bx? +ex4+d\n\nF(X...
2  M-01291  ...  Solution:\nxy =e?\n\n \n\nLength of subnormal ...
3  M-08446  ...  Solution:\n\nr=. (x) +In g(x)-2x\n\n£2 =F) +n)...
4  M-07490  ...  Solution:\nGive g(x) =x° In(x° f(x)\n\nDiffere...

[5 rows x 4 columns]
     Id_Q1    Id_Q2  Similarity_Ques  Similarity_Opt  Similarity_Sol
0  M-08446  M-08446         1.000000             1.0        1.000000
1  M-08446  M-07670         0.714286             0.0        0.312348
2  M-01289  M-00036         0.848875             1.0        0.380952
3  M-08446  M-07670         0.714286             0.0        0.312348
4  M-08363  M-07821         1.000000             1.0        0.935971
time: 1min 24s


In [ ]:
df = pd.DataFrame(columns=['Id', 'Questions', 'Options', 'Solution'])
for filename in files_B:
  temp=[]
  for i in range(len(files_B_op)):
    if files_B_op[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_B_op[i])
      for frame in gif:
          info = pytesseract.image_to_string(frame)
          temp.append(info)
  temp.sort()
  temp2 = ' '.join(temp)
  temp3=[]
  for i in range(len(files_B_sol)):
    if files_B_sol[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_B_sol[i])
      for frame in gif:
          info = pytesseract.image_to_string(frame)
          temp3.append(info)
  temp3.sort()
  temp4 = ' '.join(temp3)
  gif = imageio.get_reader(filename)
  for frame in gif:
    info = pytesseract.image_to_string(frame)
  df = df.append({'Id':filename[-14:-7],'Questions':info, 'Options':temp2, 'Solution':temp4},ignore_index=True)
df['Questions'].astype(str)
print(df.head())
similar_df_B = pd.DataFrame(columns=['Id_Q1', 'Id_Q2', 'Similarity_Ques', 'Similarity_Opt', 'Similarity_Sol'])
for k in range(df.shape[0]):
  for l in range(k+1, df.shape[0]):
     X =df.iloc[k,1]
     Y =df.iloc[l,1]
     Z =df.iloc[k,2]
     W =df.iloc[l,2]
     S1 = df.iloc[k,3]
     S2 = df.iloc[l,3]
     score1 = simi_fn(X,Y)
     score2 = simi_fn(Z,W)
     score3 = simi_fn(S1, S2)
    #print("similarity: ", cosine) 
     if score1 > 0.7:
      similar_df_B = similar_df_B.append({'Id_Q1':df.iloc[k,0], 'Id_Q2':df.iloc[l,0], 'Similarity_Ques':score1, 'Similarity_Opt':score2, 'Similarity_Sol':score3}, ignore_index=True)
print(similar_df_B) 

        Id  ... Solution
0  B-00476  ...         
1  B-00486  ...         
2  B-14749  ...         
3  B-15922  ...         
4  B-50131  ...         

[5 rows x 4 columns]
     Id_Q1    Id_Q2  Similarity_Ques  Similarity_Opt Similarity_Sol
0  B-50131  B-01388         0.737865             0.0              0
1  B-01022  B-14639         0.857143             1.0              0
2  B-01229  B-01229         1.000000             1.0              0
3  B-05343  B-05343         1.000000             1.0              0
4  B-05253  B-05253         1.000000             1.0              0
time: 1min 1s


In [ ]:
df = pd.DataFrame(columns=['Id', 'Questions', 'Options', 'Solution'])
for filename in files_C:
  temp=[]
  for i in range(len(files_C_op)):
    if files_C_op[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_C_op[i])
      for frame in gif:
          info = pytesseract.image_to_string(frame)
          temp.append(info)
  temp.sort()
  temp2 = ' '.join(temp)
  temp3=[]
  for i in range(len(files_C_sol)):
    if files_C_sol[i][-12:-7] == filename[-12:-7]:
      gif = imageio.get_reader(files_C_sol[i])
      for frame in gif:
          info = pytesseract.image_to_string(frame)
          temp3.append(info)
  temp3.sort()
  temp4 = ' '.join(temp3)
  gif = imageio.get_reader(filename)
  for frame in gif:
    info = pytesseract.image_to_string(frame)
  df = df.append({'Id':filename[-14:-7],'Questions':info, 'Options':temp2, 'Solution':temp4},ignore_index=True)
df['Questions'].astype(str)
print(df.head())
similar_df_C = pd.DataFrame(columns=['Id_Q1', 'Id_Q2', 'Similarity_Ques', 'Similarity_Opt', 'Similarity_Sol'])
for k in range(df.shape[0]):
  for l in range(k+1, df.shape[0]):
     X =df.iloc[k,1]
     Y =df.iloc[l,1]
     Z =df.iloc[k,2]
     W =df.iloc[l,2]
     S1 = df.iloc[k,3]
     S2 = df.iloc[l,3]
     score1 = simi_fn(X,Y)
     score2 = simi_fn(Z,W)
     score3 = simi_fn(S1, S2)
    #print("similarity: ", cosine) 
     if score1 > 0.7:
      similar_df_C = similar_df_C.append({'Id_Q1':df.iloc[k,0], 'Id_Q2':df.iloc[l,0], 'Similarity_Ques':score1, 'Similarity_Opt':score2, 'Similarity_Sol':score3}, ignore_index=True)
print(similar_df_C) 

        Id  ...                                           Solution
0  C-11109  ...  Solution:\n\nAt equivalent point\n\npH= tips +...
1  C-06387  ...  Solution:\n\nAt equivalent point\n\npH= ; [pKw...
2  C-06403  ...  Solution:\nLet mole fraction of O2 is x\n\n40 ...
3  C-10244  ...  Solution:\nLet mole fraction of\n\nCenters wd ...
4  C-07024  ...  Solution:\n\n \n\nd{H,0,]\n— SEES =k [Oo]!\ndt...

[5 rows x 4 columns]
     Id_Q1    Id_Q2  Similarity_Ques  Similarity_Opt  Similarity_Sol
0  C-11109  C-06387         0.764217        0.577350        0.781661
1  C-06403  C-10244         0.867149        0.866025        0.670820
2  C-06372  C-11122         0.769484        0.258199        0.524142
3  C-03481  C-11106         0.823529        0.668994        0.354787
4  C-11126  C-11127         1.000000        0.000000        0.500000
5  C-07267  C-00225         0.925476        0.833333        0.783349
6  C-11245  C-05143         0.870388        0.000000        0.735767
7  C-06076  C-10281     

In [ ]:
image_path = '/content/drive/My Drive/Duplicate_Algo_280620/Results'
count=1
for j in range(len(similar_df_P)):
  for i in range(len(files_P)):
    if files_P[i][-14:-7] == similar_df_P.iloc[j,0]:
      A = plt.imread(files_P[i])
    if files_P[i][-14:-7] == similar_df_P.iloc[j,1]:
      B = plt.imread(files_P[i])
  width = A.shape[1]
  height = A.shape[0]
  dsize = (width, height)
  output = cv2.resize(B, dsize)
  C = np.hstack((A, output))
  #cv2_imshow(C)
  cv2.imwrite(os.path.join(image_path,('P' + str(count) + '_' +  (str(similar_df_P.iloc[j,2]))[:5]+".jpg")),C)
  count+=1

time: 142 ms


In [ ]:
image_path = '/content/drive/My Drive/Duplicate_Algo_280620/Results'
count=1
for j in range(len(similar_df_M)):
  for i in range(len(files_M)):
    if files_M[i][-14:-7] == similar_df_M.iloc[j,0]:
      A = plt.imread(files_M[i])
    if files_M[i][-14:-7] == similar_df_M.iloc[j,1]:
      B = plt.imread(files_M[i])
  width = A.shape[1]
  height = A.shape[0]
  dsize = (width, height)
  output = cv2.resize(B, dsize)
  C = np.hstack((A, output))
  #cv2_imshow(C)
  cv2.imwrite(os.path.join(image_path,('M' + str(count) + '_' +  (str(similar_df_M.iloc[j,2]))[:5]+".jpg")),C)
  count+=1

time: 48.2 ms


In [ ]:
image_path = '/content/drive/My Drive/Duplicate_Algo_280620/Results'
count=1
for j in range(len(similar_df_B)):
  for i in range(len(files_B)):
    if files_B[i][-14:-7] == similar_df_B.iloc[j,0]:
      A = plt.imread(files_B[i])
    if files_B[i][-14:-7] == similar_df_B.iloc[j,1]:
      B = plt.imread(files_B[i])
  width = A.shape[1]
  height = A.shape[0]
  dsize = (width, height)
  output = cv2.resize(B, dsize)
  C = np.hstack((A, output))
  #cv2_imshow(C)
  cv2.imwrite(os.path.join(image_path,('B' + str(count) + '_' +  (str(similar_df_B.iloc[j,2]))[:5]+".jpg")),C)
  count+=1

time: 42.9 ms


In [ ]:
image_path = '/content/drive/My Drive/Duplicate_Algo_280620/Results'
count=1
for j in range(len(similar_df_C)):
  for i in range(len(files_C)):
    if files_C[i][-14:-7] == similar_df_C.iloc[j,0]:
      A = plt.imread(files_C[i])
    if files_C[i][-14:-7] == similar_df_C.iloc[j,1]:
      B = plt.imread(files_C[i])
  width = A.shape[1]
  height = A.shape[0]
  dsize = (width, height)
  output = cv2.resize(B, dsize)
  C = np.hstack((A, output))
  #cv2_imshow(C)
  cv2.imwrite(os.path.join(image_path,('C' + str(count) + '_' +  (str(similar_df_C.iloc[j,2]))[:5]+".jpg")),C)
  count+=1

time: 60.2 ms


In [ ]:
dfs = [similar_df_P,similar_df_B,similar_df_M,similar_df_C]
result = pd.concat(dfs, ignore_index=True)
result['Similarity_Sol'] = result['Similarity_Sol'].astype(float)
result = result.round(2)
(result)
#result.to_csv('result.csv')

Id_Q1    Id_Q2  ...  Similarity_Sol          combine
0   P-08354  P-08394  ...            0.82  P-08354,P-08394
1   P-00202  P-04899  ...            0.47  P-00202,P-04899
2   P-00202  P-50193  ...            0.86  P-00202,P-50193
3   P-00202  P-50293  ...            0.90  P-00202,P-50293
4   P-00202  P-50040  ...            0.61  P-00202,P-50040
5   P-04406  P-50273  ...            0.37  P-04406,P-50273
6   P-04406  P-01118  ...            0.23  P-04406,P-01118
7   P-02491  P-02546  ...            0.99  P-02491,P-02546
8   P-01118  P-22405  ...            0.58  P-01118,P-22405
9   P-03149  P-08402  ...            0.95  P-03149,P-08402
10  P-02551  P-02504  ...            0.66  P-02551,P-02504
11  P-02551  P-50193  ...            0.68  P-02551,P-50193
12  P-02551  P-50293  ...            0.76  P-02551,P-50293
13  P-02551  P-50040  ...            0.45  P-02551,P-50040
14  P-02504  P-50193  ...            0.75  P-02504,P-50193
15  P-02504  P-50293  ...            0.75  P-02504,P-50293
16  P-02504  P-50040  ...            0.54  P-02504,P-50040
17  P-50193  P-50293  ...            0.95  P-50193,P-50293
18  P-50193  P-50040  ...            0.73  P-50193,P-50040
19  P-50293  P-50040  ...            0.68  P-50293,P-50040
20  P-50293  P-04563  ...            0.76  P-50293,P-04563
21  P-07970  P-21468  ...            0.15  P-07970,P-21468
22  B-50131  B-01388  ...            0.00              NaN
23  B-01022  B-14639  ...            0.00              NaN
24  B-01229  B-01229  ...            0.00              NaN
25  B-05343  B-05343  ...            0.00              NaN
26  B-05253  B-05253  ...            0.00              NaN
27  M-08446  M-08446  ...            1.00              NaN
28  M-08446  M-07670  ...            0.31              NaN
29  M-01289  M-00036  ...            0.38              NaN
30  M-08446  M-07670  ...            0.31              NaN
31  M-08363  M-07821  ...            0.94              NaN
32  C-11109  C-06387  ...            0.78              NaN
33  C-06403  C-10244  ...            0.67              NaN
34  C-06372  C-11122  ...            0.52              NaN
35  C-03481  C-11106  ...            0.35              NaN
36  C-11126  C-11127  ...            0.50              NaN
37  C-07267  C-00225  ...            0.78              NaN
38  C-11245  C-05143  ...            0.74              NaN
39  C-06076  C-10281  ...            0.43              NaN

[40 rows x 6 columns]

time: 42 ms
